types/wdio-intercept-service.d.ts

```diff
@@ -16,43 +16,85 @@ declare namespace WdioInterceptorService {
     statusCode: number;
   }
 
-  interface InterceptedRequest {
+  interface PendingRequest {
     url: string;
     method: HTTPMethod;
-    body: string | object;
+    body: string | null | object;
+    headers: object;
+    pending: true;
+  }
+
+  interface CompletedRequest {
+    url: string;
+    method: HTTPMethod;
+    body: string | null | object;
+    pending: false;
     headers: object;
     response: {
       headers: object;
-      body: string | object;
+      body: string | null | object;
       statusCode: number;
     };
   }
+
+  type InterceptedRequest = PendingRequest | CompletedRequest;
+
+  interface OrderingStrategy {
+    /** Whether requests are ordered by time of initiation, or fulfillment */
+    orderBy?: 'START' | 'END';
+  }
+  interface AssertionRequestOrderStrategy extends OrderingStrategy {
+    /** Whether the intercepted request order must match the order in which expectations were set */
+    inOrder?: boolean;
+  }
+
+  interface GetRequestOptions extends OrderingStrategy {
+    /** Whether pending requests will be included in the response */
+    includePending?: boolean;
+  }
+  type OnlyCompletedRequests = { includePending: false };
 }
 /**
  * Convert T to T or Promise<T> depending if `@wdio/sync` is being used or not.
  */
-type AsyncSync<T> = WebdriverIO.BrowserObject extends WebDriver.Client
+type AsyncSync<T> = WebdriverIO.Browser extends WebDriver.Client
```

Comment:
@christian-bromann I don't believe this configuration is correct -- I had turned on library type checking and it complained that the `WebdriverIO` namespace didn't have an entry for `BrowserObject`, but was fine with `Browser` instead. There is still an error for `Client` not being found in the `Webdriver` namespace.

With & without this change to the `AsyncSync` conditional type, it seems this library's types always evaluate to "sync mode" in a consuming TypeScript project:

![image](https://user-images.githubusercontent.com/20871346/144750001-4c688f53-b397-43b1-8015-e914e678a31f.png)


How might this library (and others that want to provide accurate types for custom commands) best support sync & async modes in WebdriverIO versions 5, 6, and 7 (or at minimum v6 and later, as v6 is still [officially a supported version](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#back-porting-bug-fixes))?

Reply:
> How might this library (and others that want to provide accurate types for custom commands) best support sync & async modes in WebdriverIO versions 5, 6, and 7 (or at minimum v6 and later, as v6 is still officially a supported version)?

With the help of the community. While it would be amazing to provide best support for sync (with or without typescript) I doubt that we have enough resources to do so. Therefor we are happy if someone helps us with pull requests.

Reply:
I asked since you are most familiar with the wdio types found in both V6 and V7 and should be most able to tell the community what WDIO interface hierarchy is found in both sync v6 and sync v7 and thus can be used for sync detection in a d.ts file (since we cannot do a `require('@wdio/sync')` test).